src/arch/src/aarch64/gic/dist_regs.rs

```diff
@@ -0,0 +1,209 @@
+            Error::DeviceAttribute(e, true, kvm_bindings::KVM_DEV_ARM_VGIC_GRP_DIST_REGS)
+        })?;
+    } else {
+        fd.get_device_attr(&mut gic_dist_attr).map_err(|e| {
+            Error::DeviceAttribute(e, false, kvm_bindings::KVM_DEV_ARM_VGIC_GRP_DIST_REGS)
+        })?;
+    }
+    Ok(())
+}
+
+/// Get the distributor control register.
+pub fn read_ctlr(fd: &DeviceFd) -> Result<u32> {
+    let val: u32 = 0;
+    dist_attr_access(fd, GICD_CTLR, &val, false)?;
+    Ok(val)
+}
+
+/// Set the distributor control register.
+pub fn write_ctlr(fd: &DeviceFd, val: u32) -> Result<()> {
+    dist_attr_access(fd, GICD_CTLR, &val, true)
+}
+
+fn get_interrupts_num(fd: &DeviceFd) -> Result<u32> {
+    let num_irq = 0;
+
+    let mut nr_irqs_attr = kvm_bindings::kvm_device_attr {
+        group: kvm_bindings::KVM_DEV_ARM_VGIC_GRP_NR_IRQS,
+        attr: 0,
+        addr: &num_irq as *const u32 as u64,
+        flags: 0,
+    };
+    fd.get_device_attr(&mut nr_irqs_attr).map_err(|e| {
+        Error::DeviceAttribute(e, false, kvm_bindings::KVM_DEV_ARM_VGIC_GRP_NR_IRQS)
+    })?;
+    Ok(num_irq)
+}
+
+fn compute_reg_len(fd: &DeviceFd, reg: &DistReg, base: u32) -> Result<u32> {
+    let mut end = base;
+    let num_irq = get_interrupts_num(fd)?;
+    if reg.length > 0 {
+        // This is the single type register (i.e one that is not DIST_X<n>) and for which
+        // the bpi is 0.
+        // Look in the kernel for REGISTER_DESC_WITH_LENGTH.
+        end = base + reg.length as u32;
+    }
+    if reg.bpi > 0 {
+        // This is the type of register that takes into account the number of interrupts
+        // that the model has. It is also the type of register where
+        // a register relates to multiple interrupts.
+        end = base + (reg.bpi as u32 * (num_irq - super::super::layout::IRQ_BASE) / 8);
+        if reg.bpi as u32 * (num_irq - super::super::layout::IRQ_BASE) % 8 > 0 {
+            end += REG_SIZE as u32;
+        }
+    }
+    Ok(end)
+}
+
+/// Set distributor registers of the GIC.
+pub fn set_dist_regs(fd: &DeviceFd, state: &Vec<u32>) -> Result<()> {
+    let mut idx = 0;
+
+    for dreg in VGIC_DIST_REGS {
+        let mut base = dreg.base + REG_SIZE as u32 * dreg.bpi as u32;
+        let end = compute_reg_len(fd, &dreg, base)?;
+
+        while base < end {
+            let val = state[idx];
+            dist_attr_access(fd, base, &val, true)?;
+            idx += 1;
+            base += REG_SIZE as u32;
+        }
+    }
+    Ok(())
+}
+/// Get distributor registers of the GIC.
+pub fn get_dist_regs(fd: &DeviceFd) -> Result<Vec<u32>> {
+    let mut state = Vec::new();
+
+    for dreg in VGIC_DIST_REGS {
+        let mut base = dreg.base + REG_SIZE as u32 * dreg.bpi as u32;
+        let end = compute_reg_len(fd, &dreg, base)?;
+
+        while base < end {
+            let val: u32 = 0;
+            dist_attr_access(fd, base, &val, false)?;
+            state.push(val);
```

Comment:
Nit: We can preinitialize the Vec with capacity for VGIC_DIST_REGS, and then just do `dist_attr_access(fd, base, &val[i], false)?;` . Push can lead to reallocations.

Reply:
I was thinking maybe we could do something to unify `set_dist_regs` and `get_dist_regs`. Maybe we'll revisit this as part of the improvement issues.